### PR TITLE
Add hooks to download and move translations

### PR DESCRIPTION
**Pull request by DO3B**
I added the hooks. However, I didn't found a way to dynamically add "Localization" to iOS so for the moment I'm adding them using the method documentated by Apple (https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LocalizingYourApp/LocalizingYourApp.html#//apple_ref/doc/uid/10000171i-CH5-SW1). If someone as a solution, I'll take it !

### CODE REVIEW

File: hooks/before_prepare/download_translation.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env node
+
+'use strict'
+
+var child_process = require('child_process')
```
Comment (shankari) [inline hooks/before_prepare/download_translation.js]: IIUC, at the end of this hook, the translations will be in the local `locales` directory

File: hooks/before_build/android/android_move_locales.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    if (context.opts.cordova.platforms.indexOf('android') < 0 || !fs.existsSync(localesFolder))
+        return;
+    
+    var languagesFolders = fs.readdirSync(localesFolder);
+    // It's not problematic but we will remove them to have cleaner logs.
+    var filterItems = ['.git', 'LICENSE', 'README.md']
+    languagesFolders = languagesFolders.filter(item => !filterItems.includes(item));
+    console.log(LOG_NAME + "Languages found -> " + languagesFolders);
+    languagesFolders.forEach(function (language) {
+        console.log(LOG_NAME + 'I found ' + language + ", I will now move the files.")
```
Comment (shankari) [inline hooks/before_build/android/android_move_locales.js]: nit: move -> copy

File: hooks/before_build/android/android_move_locales.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        var platformRes = path.join(context.opts.projectRoot, 'platforms/android/res');
+        var wwwi18n = path.join(context.opts.projectRoot, 'www/i18n/');
+        var languageFolder = localesFolder + "/" + language;
+
+        var values = "/values-" + language;
+        var valuesFolder = path.join(languageFolder, values);
+        if (fs.existsSync(valuesFolder)) {
+            console.log(LOG_NAME + "Moving " + valuesFolder + " to " + platformRes);
```
Comment (shankari) [inline hooks/before_build/android/android_move_locales.js]: Ditto.

File: hooks/before_build/android/android_move_locales.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+            fs.copySync(valuesFolder, platformValues);
+            console.log(LOG_NAME + valuesFolder + "copied...")
+        } else {
+            console.log(LOG_NAME + valuesFolder + " not found, I will continue.")
+        }
+
+        var languagei18n = path.join(languageFolder, "/i18n/");
+        if (fs.existsSync(languagei18n)) {
+            console.log(LOG_NAME + "Moving " + languagei18n + " to " + wwwi18n);
```
Comment (shankari) [inline hooks/before_build/android/android_move_locales.js]: Ditto

File: hooks/after_platform_add/ios/ios_move_locales.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        languagesFolders.forEach(function (language) {
+            console.log(LOG_NAME + 'I found ' + language + ", I will now move the files.")
+            var platformRes = path.join(context.opts.projectRoot, 'platforms/ios/' + applicationName + "/Resources/");
+            var wwwi18n = path.join(context.opts.projectRoot, 'www/i18n/');
+            var languageFolder = localesFolder + "/" + language;
+
+            var lproj = "/" + language + ".lproj";
+            var lprojFolder = path.join(languageFolder, lproj);
+            if (fs.existsSync(lprojFolder)) {
+                console.log(LOG_NAME + "Moving " + lprojFolder + " to " + platformRes);
```
Comment (shankari) [inline hooks/after_platform_add/ios/ios_move_locales.js]: Nit: Moving -> Copying

File: hooks/after_platform_add/ios/ios_move_locales.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+                fs.copySync(lprojFolder, platformlproj);
+                console.log(LOG_NAME + lprojFolder + "copied...")
+            } else {
+                console.log(LOG_NAME + lprojFolder + " not found, I will continue.")
+            }
+
+            var languagei18n = path.join(languageFolder, "/i18n/");
+            if (fs.existsSync(languagei18n)) {
+                console.log(LOG_NAME + "Moving " + languagei18n + " to " + wwwi18n);
```
Comment (shankari) [inline hooks/after_platform_add/ios/ios_move_locales.js]: Ditto

File: hooks/after_platform_add/ios/ios_move_locales.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            if (fs.existsSync(lprojFolder)) {
+                console.log(LOG_NAME + "Moving " + lprojFolder + " to " + platformRes);
+
+                var platformlproj = platformRes + lproj;
+                if (!fs.existsSync(platformlproj)) {
+                    console.log(LOG_NAME + platformlproj + "does not exist, I will create it.");
+                    fs.mkdirSync(platformlproj);
+                }
+
+                fs.copySync(lprojFolder, platformlproj);
```
Comment (shankari) [inline hooks/after_platform_add/ios/ios_move_locales.js]: So I am not quite sure what you mean by "I didn't found a way to dynamically add "Localization" to iOS ". What is not working? 

Are you worried about the "Base Internationalization"? It looks like that is mainly focused around automatically extracting user-facing strings from 

> Base internationalization separates user-facing strings from .storyboard and .xib files. It relieves localizers of the need to modify .storyboard and .xib files in Interface Builder....

We don't really have complex storyboards or xib files - the storyboard just uses a webview and the rest of the UX is HTML + CSS + JS.

The only user visible strings are the ones that you pulled out into the strings files. So you really need the "Creating Strings Files for User-Facing Text in Your Code" section of [maintaining your own strings files](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/MaintaingYourOwnStringsFiles/MaintaingYourOwnStringsFiles.html#//apple_ref/doc/uid/10000171i-CH19-SW22)

and that basically just outlines this procedure, with the caveat that you can use `genstrings` if needed to autogenerate the strings file by looking at `NSLocalizedString`. But you have already manually generated this so I think you are good.

If you have more details on what exactly you are worried about, I can see if I can suggest something.

File: hooks/before_prepare/download_translation.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    // Checking if git is installed, return error if not. 
+    try {
+        child_process.execSync('which git');
+    } catch (err) {
+        console.error(LOG_NAME + 'git not found, (' + err + ')');
+        return;
+    }
+
+    if (!fs.existsSync(localesFolder)) {
+        child_process.execSync('git clone https://github.com/e-mission/e-mission-translate ' + localesFolder, { 'timeout': 10000 });
```
[the remark this answers is not in the document]
Comment (DO3B) [reply hooks/before_prepare/download_translation.js]: I added the {'stdio':'inherit'} (https://github.com/e-mission/e-mission-phone/pull/593/commits/aa52853983601420446b4b4aca27793da2a1284e)

File: hooks/before_prepare/download_translation.js (hunk 1 above)
Comment (DO3B) [reply hooks/before_prepare/download_translation.js]: Yes, and if the directory already exists, it will just pull the repo instead of cloning it.

**Pull request merged**